module/zfs/dsl_scan.c

```diff
@@ -1833,6 +1833,33 @@ dsl_scan_scrub_done(zio_t *zio)
 	mutex_exit(&spa->spa_scrub_lock);
 }
 
+static boolean_t
+dsl_scan_need_resilver(spa_t *spa, const dva_t *dva, size_t size, uint64_t phys_birth)
+{
+	vdev_t *vd;
+
+	if (DVA_GET_GANG(dva)) {
+		/*
+		 * Gang members may be spread across multiple
+		 * vdevs, so the best estimate we have is the
+		 * scrub range, which has already been checked.
+		 * XXX -- it would be better to change our
+		 * allocation policy to ensure that all
+		 * gang members reside on the same vdev.
+		 */
+		return (B_TRUE);
+	}
+
+	vd = vdev_lookup_top(spa, DVA_GET_VDEV(dva));
+	if (!vdev_dtl_contains(vd, DTL_PARTIAL, phys_birth, 1))
+		return (B_FALSE);
+
+	if (vd->vdev_ops == &vdev_raidz_ops)
```

Comment:
By doing `vdev_dtl_contains(vd, DTL_PARTIAL, phys_birth, 1)` first, I was able to implement several optimizations in `vdev_raidz_need_resilver()`, e.g. at the beginning:
```
if (s + nparity >= dcols) /* spans all child vdevs */
	return (B_TRUE);
```
The whole function was essential skipped for the big blocks. This is correct only under the assumption that `vdev_dtl_contains(vd, DTL_PARTIAL, phys_birth, 1)` is true.